torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -1147,7 +1147,35 @@ void InitXlaModuleBindings(py::module m) {
   m.def("_xla_get_tensor_id",
         [](const at::Tensor& tensor) { return GetTensorId(tensor); });
   m.def("_xla_get_spmd_config_is_locked", []() { return GetLockSpmdConfig(); });
+  m.def("_xla_force_spmd_device", []() {
+    // It is actually more easier to force SPMD mode than blocking if there is
+    // non-SPMD initialized tensors, for the 3rd-party solution integration. For
+    // instance, HuggingFace trainer pre-loads embeddings table and the training
+    // initialization is done over multiple scripts. Being able to force SPMD
+    // allows the users to call `xr.use_spmd()` more freely, given that the
+    // earlier they call, the smaller the one-time overhead of replicating
+    // non-SPMD backed tensors.
+    torch::lazy::BackendDevice backend_device = bridge::GetCurrentDevice();
+    std::vector<XLATensorPtr> xtensors =
+        XLAGraphExecutor::Get()->GetLiveTensors(&backend_device);
+    for (auto xtensor : xtensors) {
+      XlaDeviceType xla_device_type =
+          static_cast<XlaDeviceType>(xtensor->GetDevice().type());
+      if (xla_device_type != XlaDeviceType::SPMD) {
+        // TODO(yeounoh) this can be further optimized via CopyToDevice.
+        at::Tensor tensor = xtensor->ToTensor(false);
+        xtensor->SetXlaData(
+            TensorToXlaData(tensor, ParseDeviceString("SPMD:0")));
+      }
```

Comment:
We statically track whether to execute in SPMD mode here: https://github.com/pytorch/xla/blob/13dc9df4df52225f428e059f54661d1f9481fd65/torch_xla/csrc/device.cpp#L86

It's initialized on the first `UseVirtualDevice` call. Not sure if it's an issue, but could we `XLA_CHECK(UseVirtualDevice())` at the end to ensure it's being set properly?

Reply:
Oh, that's set outside `force_spmd_device` but in `use_spmd`. Bascially `UseVirtualDevice` looks at `XLA_USE_SPMD` envvar, but in C++ we cannot update the envvar system-wide.

Reply:
Sorry if I'm missing something 😅 

Based on https://github.com/pytorch/xla/blob/aa0641467b335f22a00d97f33ae8be6049e81624/torch_xla/runtime.py#L245-L251, we only call `force_spmd_device` when the config is locked, so `UseVirtualDevice`'s return value has been fixed (to `false` without the env var set). So would the runtime would still expect an unsharded computation?

Here's a program which demonstrates the problem:

```python
import torch_xla.runtime as xr, torch_xla, torch

print(xr.addressable_device_count()) # Calls into `_xla_num_devices` which locks the config
t = torch.randn(2, 2).to('xla')  # Data is on non-SPMD device
xr.use_spmd()  # Now we convert the data to replicated for SPMD

print(t + t) # Does this execute replicated?
```

Reply:
Ahhhhh, I get it now `static bool` -- "it's initialized on the first call" good catch. Will look into it.

Reply:
@jonb377 do you know if that needs to be static, or we can just do the envvar lookup? I don't have the context as to why that was declared static. cc @JackCaoG

Reply:
hm my guess is that's because we can't set envvar system-wide in python... let me work around it.

Reply:
Resolved, and updated the unit test inspired by @jonb377 's example ❤️ 
```
...

    # execute replicated
    self.assertTrue(
        "{replicated}" in torch_xla._XLAC._get_xla_tensors_hlo([t + t]))
```

Reply:
Hmm I think the static bool is just an optimization to avoid querying the environment on each lookup. I'm not sure how significant the overhead would be without it.